Allow top-level libraries in `old_public_name`

nojb opened the pull request:
This PR proposes to extend `(deprecated_library name ...)` to allow `old_public_name` to be a top-level library that is not a sub-package of the main package. While it is not possible to define such libraries with `dune`, it is possible to do so using other build systems (e.g. `menhir` does this). The intent of this PR is to help migration of such projects to `dune`.

The syntax to redirect a top-level library `foo` belonging to a package `p` to `bar` is
```
(deprecated_library_name
 (old_public_name foo)
 (package p)
 (new_public_name bar))
```
Suggestions for a better syntax are welcome!

**There is one subtlety though.** Suppose by way of example that we have a legacy project `p` with a top-level library `mylib` (main module: `MyLib`) that we are porting to `dune`. Naturally we will now give the library the name `mylib` (since it has to match the name of the main module) and public name `p.mylib`. 
```
(library
 (name mylib)
 (public_name p.mylib))
```
In order to maintain backwards compatibility we would like to do:
```
(deprecated_library_name
 (old_public_name mylib)
 (package p)
 (new_public_name p.mylib))
```
The problem is that `dune` will complain that there are two definitions of the library `mylib` (one for each stanza).

The last commit ("Fix (?) ...") tries to fix this issue, it seems to work fine, but am not 100% sure about its implications and would appreciate a second opinion.

rgrinberg commented:
I haven't looked at the implementation yet, but it seems a bit fishy to expose this `package` argument.

Ideally, the deprecated stanza should belong to both packages if they exist in the dune-project.

diml commented:
@rgrinberg do you mean that the deprecated stanza would end up in both `dune-package` files? Putting the deprecated `menhirLib` name in the `dune-package` file of `menhir` wouldn't work since dune wouldn't eagerly go and look in `<libdir>/menhir/dune-package` to resolve the `menhirLib` name.

> The problem is that dune will complain that there are two definitions of the library mylib (one for each stanza).

At the user level, there is no ambiguity. Whether we interpret `menhirLib` as the deprecated name for `menhir.lib` or the local name of `menhir.lib`, both point to the exact same library. We should detect such cases and simply accept them.

rgrinberg commented:
Duplicating the stanza for both of the packages that it touches is what I had in mind. It’s not exactly in elegant as there are cases where dune cannot do it. It’s not possible when the packages aren’t in the same project for example. It still seems like a useful addition.

> do you mean that the deprecated stanza would end up in both dune-package files? Putting the deprecated menhirLib name in the dune-package file of menhir wouldn't work since dune wouldn't eagerly go and look in <libdir>/menhir/dune-package to resolve the menhirLib name.

diml commented:
What is point of putting the old name in the new package?

nojb commented:
> What is point of putting the old name in the new package?

Not sure if I understand the question, but in the new package the "name" (not the public name) needs to be the same as before because it need to match the name of the main module. The "public name" could be different of course.

nojb committed:
Add test

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
> At the user level, there is no ambiguity. Whether we interpret `menhirLib` as the deprecated name for `menhir.lib` or the local name of `menhir.lib`, both point to the exact same library. We should detect such cases and simply accept them.

Good point. I pushed a commit implementing this extra check ("Handle top-level libs..."). I am not very happy with the implementation, suggestions welcome.

Also, I think a better syntax is needed.

diml commented:
> Not sure if I understand the question, but in the new package the "name" (not the public name) needs to be the same as before because it need to match the name of the main module. The "public name" could be different of course.

Agreed. My question was not very clear, it was in response to:

> Duplicating the stanza for both of the packages that it touches is what I had in mind. It’s not exactly in elegant as there are cases where dune cannot do it. It’s not possible when the packages aren’t in the same project for example. It still seems like a useful addition.

> Good point. I pushed a commit implementing this extra check ("Handle top-level libs..."). I am not very happy with the implementation, suggestions welcome.

What about replacing the `Lib_name.Map.of_list` by `Lib_name.Map.of_list_reduce` and dealing with duplicated names there?

> Also, I think a better syntax is needed.

Here are a few suggestions:

- `(old_public_name menhirLib in package menhir)`
- `(old_public_name menhirLib (package menhir))`

BTW, I was thinking again about what we were talking about the other day on slack, and my gut feeling is that things will work better in the future if all toplevel names appear in the `dune-project` file. For instance, we could add a field in the `package` stanza to declare all the toplevel names:

```scheme
(package
 (name menhir)
 (claim_deprecated_package_names menhirLib))
```

to indicate that `menhir` claims the `menhirLib` package name. Then we would only be allowed to use an `old_public_name` that is part of a package names mentioned in the dune-project file. We might not even need the `(package ...)` field anymore in the `deprecated_library_name` stanza.

nojb commented:
> BTW, I was thinking again about what we were talking about the other day on slack, and my gut feeling is that things will work better in the future if all toplevel names appear in the `dune-project` file. For instance, we could add a field in the `package` stanza to declare all the toplevel names:
> 
> ```scheme
> (package
>  (name menhir)
>  (claim_deprecated_package_names menhirLib))
> ```
> 
> to indicate that `menhir` claims the `menhirLib` package name. Then we would only be allowed to use an `old_public_name` that is part of a package names mentioned in the dune-project file. We might not even need the `(package ...)` field anymore in the `deprecated_library_name` stanza.

Indeed, it sounds like a better place to put this information; I will give it a try. Thanks!

nojb committed:
Add (deprecated_package_names ...) field

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Add Lib.Resolve_result.to_dyn

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Update CHANGES.md

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Update manual

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
I reworked this PR.

Now there is a new field `(deprecated_package_names pkg1 pkg2 ...)` in the `(package (name p) ...)` stanza in `dune-project`. The names `pkg1`, `pkg2`, ... can be used in the `(old_public_name ...)` field of the `(deprecated_library_name ...)` stanza. These names are interpreted as top-level libraries attached to `p`. Appropriate `META` files are generated.

I refactored a bit the code checking for duplicate library definitions to handle the fact that the same name can appear in the `(deprecated_package_names ...)` field and in the `(name ...)` field of a library.

Review welcome. Thanks!

diml commented on src/dune/dune_file.ml:
```diff
@@ -630,27 +630,47 @@ module Public_lib = struct
-  let make project ((_, s) as loc_name) =
+  let package t = t.package
+
+  let make ?allow_deprecated_names project ((_, s) as loc_name) =
```
We have been using the `?(allow_deprecated_names=false)` form in Dune so far, let's stick to that for consistency.

nojb replied:
Fixed.

diml commented on src/dune/dune_file.ml:
```diff
@@ -2191,23 +2211,29 @@ module Deprecated_library_name = struct
     ; project : Dune_project.t
-    ; old_public_name : Public_lib.t
-    ; new_public_name : Lib_name.t
+    ; old_public_name : bool * Public_lib.t
```
It's not obvious what this `bool` mean without reading the parsing code. We should use a dedicated type or a record.

nojb replied:
Fixed, I defined a new record type.

diml commented on src/dune/package.mli:
```diff
@@ -81,6 +81,7 @@ type t =
   ; version : string option
   ; kind : Kind.t
   ; tags : string list
+  ; deprecated_package_names : (Loc.t * string) list
```
I'd use `Loc.t Name.Map.t` here

nojb replied:
Fixed.

diml commented:
Looks good. We also need to generate a `dune-package` file for deprecated package names, otherwise deprecated library names won't translate to `Redirect` when used from another project, and so this would break `(implicit_transitive_deps false)`.

BTW, this PR doesn't allow `(old_public_name <deprecated-name>.<something>)`. Would that be useful to support?

diml commented on src/dune/install_rules.ml:
```diff
@@ -170,17 +174,37 @@ end = struct
                let dune_package_file =
                  Package_paths.dune_package_file ctx pkg
                in
+               let toplevel_lib_meta_files =
```
It seems to me that we could use `pkg.deprecatd_package_names` here

nojb replied:
We could, but right now we don't check that every name mentioned in `(deprecated_package_names)` actually appears in a `deprecated_library_name` stanza. Should we enforce that?

diml replied:
This is similar to checking that packages are empty, which we don't do, so we shouldn't do it here either.

nojb commented:
> Looks good. We also need to generate a `dune-package` file for deprecated package names, otherwise deprecated library names won't translate to `Redirect` when used from another project, and so this would break `(implicit_transitive_deps false)`.

I am not very familiar with `dune-package` files; should this file be generated in the same directory as the META file?

diml commented:
Yep. Basically, `dune-package` are dune's version of META files. They are in s-expression syntax and store more information. They are not meant to be read by tools other than `dune`.

nojb commented:
> Yep. Basically, `dune-package` are dune's version of META files. They are in s-expression syntax and store more information. They are not meant to be read by tools other than `dune`.

OK; is it correct that the redirection entry for the deprecated package should not appear then in the `dune-package` file of the main package ?

diml commented:
Yes, that's correct. For instance, for menhir, the `(deprecated_library_name (old_public_name menhirLib) (new_public_name menhir.lib))` stanza should appear only in `<libdir>/menhirLib/dune-package`. Same for the `META` file.

nojb commented:
> Yes, that's correct. For instance, for menhir, the `(deprecated_library_name (old_public_name menhirLib) (new_public_name menhir.lib))` stanza should appear only in `<libdir>/menhirLib/dune-package`. Same for the `META` file.

Got it, thanks. One more question : the `name` in the `dune-package` file for the deprecated library should be the name of the deprecated library (eg `menhirLib`) or of the main package (eg `menhir`) ?

nojb committed:
Add Dune_package.{to_dyn,Or_meta.pp}

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

diml commented:
it should be `menhirLib`. In fact, it's basically redundant with the directory name.

nojb commented:
> it should be `menhirLib`. In fact, it's basically redundant with the directory name.

OK. I added the generation of `dune-package` files, and I think all other issues have been addressed.

diml commented on src/dune/install_rules.ml:
```diff
@@ -170,17 +178,45 @@ end = struct
+                           ~dst:(Package.Name.to_string name ^ "/" ^ Dune_package.fn)) ::
+                       []
+                     | _ -> [])
+                   (Super_context.lib_entries_of_package sctx pkg.name)
```
Thinking more about this, we should really use `pkg.deprecated_package_names`. It will be more efficient. The only difference is that we would generate a `META` and `dune-package` file  even if the deprecated packages are empty, but that actually seems like a good thing to me. In particular, this is in line with normal packages.

nojb replied:
Makes sense, I will change it.

nojb replied:
This is now done.

diml commented:
Regarding allowing `<deprecated-name>.<sometihng>`, the more I think about it and the more I think we should allow it. It doesn't seem very complicated to support, and if someone needs it in the future we won't need more versioning. What do you think?

nojb commented:
> Regarding allowing `<deprecated-name>.<sometihng>`, the more I think about it and the more I think we should allow it. It doesn't seem very complicated to support, and if someone needs it in the future we won't need more versioning. What do you think?

Sorry I missed this comment before. Yes, it makes sense, I will look into it.

nojb commented:
> > Regarding allowing `<deprecated-name>.<sometihng>`, the more I think about it and the more I think we should allow it. It doesn't seem very complicated to support, and if someone needs it in the future we won't need more versioning. What do you think?
> 

This is now working. I added some tests as an example.

nojb commented on src/dune/install_rules.ml:
```diff
@@ -262,6 +293,18 @@ let gen_dune_package sctx pkg =
   let dune_version =
     Dune_lang.Syntax.greatest_supported_version Stanza.syntax
   in
+  let lib_entries = Super_context.lib_entries_of_package sctx pkg.name in
```
This call used to be delayed under `gen_dune_package ()` below, we know call it eagerly. Probably harmless, but just pointing it out.

diml replied:
Thanks for point that out. That seems ok indeed

diml commented:
One possible follow-up: currently we can use `(package x)` as a dependency. However if `x` is a deprecated package present in the workspace, I don't believe it will work. It should be possible to setup a redirection when we setup the alias used to interpret `(package ...)` in `install_rules` in `install_rules.ml`.

nojb commented:
@diml looking at what we did once more, it seems to me it would make more sense to have the stanza `(deprecated_package_names ...)` at the toplevel of `dune-project` instead of being a subfield of `(package ...)`. After all, there isn't anything tying the deprecated packages to the "parent" package.

nojb commented:
(I think this would be a change of syntax only, the implementation already considers deprecated package names to be globally defined)

diml commented:
Well, if you write `(package (name foo) (deprecated_package_names bar))`, then the files of `bar` such as `bar/META` will be installed as part of package `foo`, right? Otherwise, we'd need a separate opam entry for `bar`

diml commented:
Well, if you write `(package (name foo) (deprecated_package_names bar))`, then the files of `bar` such as `bar/META` will be installed as part of package `foo`, right? Otherwise, we'd need a separate opam entry for `bar`

nojb commented:
> Well, if you write `(package (name foo) (deprecated_package_names bar))`, then the files of `bar` such as `bar/META` will be installed as part of package `foo`, right? Otherwise, we'd need a separate opam entry for `bar`

Yes you are right, I forgot we need to associate deprecated packages with the package they need to be installed by. Another possibility is to use:
```
(package (name foo) (deprecated_by bar))
```
for a deprecated name `foo`. But I am also OK leaving things as they are.

diml commented:
That seems like a valid choice, but I feel like the choice of syntax should reflect the behaviour. More precisely:

1. if the deprecated names are are a field of the main package, then the files from deprecated packages are just part of the main package and there is no separate opam package
2. if the deprecated package has its own top-level stanza, then the deprecated package should have its own `<package>.opam` file, generate its own `<package>.install` file and ultimately have its own entry in opam

Given that the use case for this feature is for project for which the ocamlfind package name doesn't match the opam package name, I feel like option 1 is better.

nojb commented:
> Given that the use case for this feature is for project for which the ocamlfind package name doesn't match the opam package name, I feel like option 1 is better.

Makes sense. I added a couple of more tests and rebased to clean up the history. Unless someone objects I am planning to merge today after CI pases.

nojb committed:
make fmt

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>
